File: src/main/java/com/fendyk/clients/FetchAPI.java
package com.fendyk.clients;

import com.fendyk.DTOs.MinecraftUserDTO;
import com.fendyk.utilities.Log;
import com.fendyk.Main;
import com.google.gson.JsonElement;
import com.google.gson.JsonNull;
import com.google.gson.JsonParser;
import okhttp3.*;
import org.bukkit.Bukkit;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.Arrays;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

public abstract class FetchAPI<DTO, UpdateDTO> {
    static Main main = Main.getInstance();
    protected static OkHttpClient client = new OkHttpClient.Builder()
            .readTimeout(30, TimeUnit.SECONDS)
            .connectTimeout(30, TimeUnit.SECONDS)
            .build();
    protected static final MediaType JSON = MediaType.parse("application/json; charset=utf-8");
    protected static Request.Builder requestBuilder;
    private static String url;

    private final Class<DTO> dtoType;
    private final String endpointUrl;

    public FetchAPI(String endpointUrl, Class<DTO> dtoType) {
        this.endpointUrl = endpointUrl;
        this.dtoType = dtoType;
    }

    /**
     * Fetches from api
     *
     * @param request
     * @param name
     * @return
     */
    @Nullable
    protected CompletableFuture<JsonElement> fetch(Request request) {
        final boolean inDebugMode = main.getServerConfig().isInDebugMode();
        return CompletableFuture.supplyAsync(() -> {
            try (Response response = client.newCall(request).execute()) {
                if (inDebugMode) {
                    Log.info("");
                    Log.info("Request URL: " + request.url());
                    Log.info("Response Code: " + response.code());
                    Log.info("Response Message: " + response.message());
                    Log.info("");
                }

                if (!response.isSuccessful()) {
                    throw new IOException("Unexpected response code: " + response.code() + response.message());
                }

                try (ResponseBody body = response.body()) {
                    if (body == null) throw new IOException("Could not find a body.");
                    String json = body.string();
                    if (inDebugMode) {
                        Log.info("Response Body: " + json);
                    }
                    return JsonParser.parseString(json);
                }
            } catch (IOException e) {
                Log.error("Fetch error: " + e.getMessage());
                Log.error("Request url: " + request.url());
                Log.error("Stacktrace: " + Arrays.toString(e.getStackTrace()));
                return null;
            }
        });
    }

    public CompletableFuture<DTO> get(String key) {
        final String finalUrl = endpointUrl + "/" + key;
        return CompletableFuture.supplyAsync(() -> {
            try {
                Request request = requestBuilder
                        .url(url + finalUrl)
                        .build();
                CompletableFuture<JsonElement> aFetch = fetch(request);
                if(aFetch == null) throw new Exception("Fetching failed with GET: " + finalUrl);
                return Main.gson.fromJson(
                        aFetch.join(),
                        dtoType
                );
            }
            catch (Exception e) {
                return null;
            }
        });
    }

    public CompletableFuture<DTO> create(DTO dto) {
        RequestBody body = RequestBody.create(Main.gson.toJson(dto), JSON);
        return CompletableFuture.supplyAsync(() -> {
            try {
                Request request = requestBuilder
                        .url(url + endpointUrl)
                        .post(body)
                        .build();
                CompletableFuture<JsonElement> aFetch = fetch(request);
                if(aFetch == null) throw new Exception("Fetching failed with CREATE: " + endpointUrl);
                return Main.gson.fromJson(
                        aFetch.join(),
                        dtoType
                );
            }
            catch (Exception e) {
                return null;
            }
        });
    }

    public CompletableFuture<DTO> update(String key, UpdateDTO dto) {
        final String finalUrl = endpointUrl + "/" + key;
        RequestBody body = RequestBody.create(Main.gson.toJson(dto), JSON);
        return CompletableFuture.supplyAsync(() -> {
            try {
                Request request = requestBuilder
                        .url(url + finalUrl)
                        .post(body)
                        .build();
                CompletableFuture<JsonElement> aFetch = fetch(request);
                if(aFetch == null) throw new Exception("Fetching failed with UPDATE: " + finalUrl);
                return Main.gson.fromJson(
                        aFetch.join(),
                        dtoType
                );
            }
            catch (Exception e) {
                return null;
            }
        });
    }

    public CompletableFuture<DTO> delete(String key) {
        final String finalUrl = endpointUrl + "/" + key;
        return CompletableFuture.supplyAsync(() -> {
            try {
                Request request = requestBuilder
                        .url(url + finalUrl)
                        .delete()
                        .build();
                CompletableFuture<JsonElement> aFetch = fetch(request);
                if(aFetch == null) throw new Exception("Fetching failed with DELETE" + finalUrl);
                return Main.gson.fromJson(
                        aFetch.join(),
                        dtoType
                );
            }
            catch (Exception e) {
                return null;
            }
        });
    }

    public static void connect(String url, String jwtKey) {
        FetchAPI.url = url;
        FetchAPI.requestBuilder = new Request.Builder()
                .addHeader("Authorization", "Bearer " + jwtKey);
    }

}

File: src/main/java/com/fendyk/clients/FetchAPI.kt
package com.fendyk.clients

import com.fendyk.Main
import com.fendyk.utilities.Log
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeUnit

abstract class FetchAPI<K, DTO, UpdateDTO>(private val dtoType: Class<DTO>) {
    enum class RequestMethod(val method: String) {
        GET("GET"),
        POST("POST"),
        PATCH("PATCH"),
        DELETE("DELETE")
    }

    fun fetch(url: String, requestMethod: RequestMethod, data: Any?): CompletableFuture<DTO> {
        val body: RequestBody? = if (data != null) Main.gson.toJson(data).toRequestBody(JSON) else null
        val request: Request = requestBuilder
                .url(Companion.url + "/minecraftusers/" + url)
                .method(requestMethod.toString(), body)
                .build()
        return CompletableFuture.supplyAsync {
            try {
                client.newCall(request).execute().use { response ->
                    if (main.serverConfig.isInDebugMode) {
                        Log.info("")
                        Log.info("Request URL: " + request.url)
                        Log.info("Response Code: " + response.code)
                        Log.info("Response Message: " + response.message)
                        Log.info("")
                    }
                    if (!response.isSuccessful) {
                        throw IOException("Unexpected response code: " + response.code)
                    }
                    response.body.use { responseBody ->
                        if (responseBody == null) return@supplyAsync null
                        val json = responseBody.string()
                        if (main.serverConfig.isInDebugMode) {
                            Log.info("Response Body: $json")
                        }
                        return@supplyAsync Main.gson.fromJson<DTO>(
                                json,
                                dtoType
                        )
                    }
                }
            } catch (e: IOException) {
                Log.error("Error in fetchFromApi: " + e.message)
                Log.error("Stacktrace: " + Arrays.toString(e.stackTrace))
                return@supplyAsync null
            }
        }
    }

    open fun get(key: K): CompletableFuture<DTO> {
        val future = CompletableFuture<DTO>()
        future.completeExceptionally(Exception("GET is not implemented yet."))
        return future;
    }
    open fun create(dto: DTO): CompletableFuture<DTO> {
        val future = CompletableFuture<DTO>()
        future.completeExceptionally(Exception("CREATE is not implemented yet."))
        return future;
    }
    open fun update(key: K, dto: UpdateDTO): CompletableFuture<DTO> {
        val future = CompletableFuture<DTO>()
        future.completeExceptionally(Exception("UPDATE is not implemented yet."))
        return future;
    }
    open fun delete(key: K): CompletableFuture<DTO> {
        val future = CompletableFuture<DTO>()
        future.completeExceptionally(Exception("DELETE is not implemented yet."))
        return future;
    }


    companion object {
        var main: Main = Main.getInstance()
        protected var client: OkHttpClient = OkHttpClient.Builder()
                .readTimeout(10, TimeUnit.SECONDS)
                .connectTimeout(10, TimeUnit.SECONDS)
                .build()
        @JvmField
        protected val JSON: MediaType? = ("application/json; charset=utf-8").toMediaTypeOrNull()
        protected lateinit var url: String;
        protected const val authHeader: String = "Authorization"
        @JvmField
        protected var requestBuilder: Request.Builder = Request.Builder()
        @JvmStatic
        fun connect(url: String, jwtKey: String) {
            FetchAPI.url = url;
            requestBuilder.addHeader(authHeader, "Bearer $jwtKey")
        }
    }
}
